_drgn.pyi

```diff
@@ -448,7 +450,39 @@ class Program:
         :param fn: Callable taking a program, name, :class:`FindObjectFlags`,
             and filename: ``(prog, name, flags, filename)``. The filename
             should be matched with :func:`filename_matches()`. This should
-            return an :class:`Object`.
+            return an :class:`Object` or ``None`` if not found.
+        """
+        ...
+    def add_symbol_finder(
+        self, fn: Callable[[Optional[str], Optional[int], bool], List[Symbol]]
+    ) -> None:
+        """
+        Register a callback for finding symbols in the program.
+
+        The callback should take three arguments: a search name, a search
+        address, and a boolean flag indicating whether to return the first
+        match. When the flag is True, the callback should return just one
+        :class:`Symbol`. When the flag is False, the callback should return a
+        list of all matching :class:`Symbol`\\ s. Both the name and address
+        arguments are optional. If both are provided, then the result(s) should
+        match both. If neither are provided the finder should return all
+        available symbols.
+
+        Callbacks are called in reverse order of the order they were added. When
+        the boolean flag is set, the search will short-circuit after the first
+        finder returns a result. Otherwise, all callbacks will be called, and
+        all results will be returned.
+
+        .. note::
+
+            There is one case where callback order is not respected: drgn's
+            internal symbol resolution. If drgn has debuginfo already loaded for
+            a particular memory address, it will take a shortcut and directly
+            use the built-in ELF symbol finder first. If this fails, it will
+            continue to call each callback in the order described.
+
```

Comment:
This note is outdated. I removed it for the second revision, but it shouldn't have even been included in the first revision. If you look at `drgn_find_symbol_by_address_internal()`, it also uses the symbol finder callbacks.

I wrote this note before I had incorporated a `struct drgn_module *` into the symbol finder API. Without that pointer, it was impossible to implement the shortcut for address lookup with `dwfl_addrmodule()`. So I ended up leaving `drgn_find_symbol_by_address_internal()` to use the internal ELF symbol code. But thankfully that's all gone now: the only difference between the internal and external one as that the internal one does not return a lookup error.